jsdoc.el

```diff
@@ -116,67 +117,96 @@
-    (_ "any")))
+  (pcase (treesit-node-type node)
+    ("identifier" (jsdoc--infer-identifier node))
+    ("true" "boolean")
+    ("false" "boolean")
+    ("number" "number")
+    ("string" "string")
+    ("array" "*[]")
+    ("object" "object")
+    ("new_expression" (jsdoc--infer-type (treesit-node-child node 0 t)))
+    ("call_expression" (jsdoc--infer-type (treesit-node-child node 0 t)))
+    ("arrow_function" (jsdoc--infer-closure-type node))
+    ("binary_expression" (jsdoc--infer-binary-expression node))
+    (_ "*")))
+
+(defun jsdoc--infer-closure-type (node)
+  "Return the inferred type of the given closure NODE."
+  (let* ((fn (jsdoc--parse-generic-function node ""))
+         (params (plist-get fn :params))
+
+         ;; Reduce the list of params if more than 1
+         (list (if (< 1 (length params))
+                   (--reduce (format "%s, %s"
+                                     (or (plist-get acc :type)
+                                         acc)
+                                     (plist-get it :type))
+                             params)
+                 (plist-get (car params) :type)
+                 ))
+         (returns (plist-get fn :returns)))
+    (concat "function(" list "): " returns)))
+
+(defun jsdoc--infer-array-type (node)
+  "Return the inferred type of the given array NODE."
+  (let* ((params (--reduce
+                  (if (eq acc it)
+                      acc
+                    (s-concat acc "|" it))
+                  (--map (plist-get (jsdoc--parse-param it) :type) node))))
+    (if (string-match-p (regexp-quote "|") params)
+        (concat "(" params ")" "[]")
+      (concat params "[]"))))
 
 (defun jsdoc--infer-binary-expression (node)
   (format "TODO"))
 
 (defun jsdoc--infer-identifier (node)
-  "Return given identifier NODE type.  `X' if `X()', otherwise `any'."
-  (let* ((next-sibling (tsc-get-next-named-sibling node)))
+  "Return given identifier NODE type.  `X' if `X()', otherwise `*'."
+  (let* ((next-sibling (treesit-node-next-sibling node t)))
     (if (and next-sibling
-             (equal (tsc-node-type next-sibling) 'arguments)
-             (s-uppercase? (substring (tsc-node-text node) 0 1)))
-        (tsc-node-text node)
-      "any")))
+             (equal (treesit-node-type next-sibling) 'arguments)
+             (s-uppercase? (substring (treesit-node-text node) 0 1)))
+        (treesit-node-text node)
+      "*")))
 
 (defun jsdoc--get-return-type (node)
   "Return the return type of given NODE."
   (-if-let (return-type (jsdoc--get-returned-type-of-statement node 'return_statement))
-      (pcase (tsc-node-text (tsc-get-nth-child node 0))
+      (pcase (treesit-node-text (treesit-node-child node 0))
         ("async" (format "Promise<%s>" return-type))
         (_ return-type))
     (progn
-      (jsdoc--infer-type (tsc-get-child-by-field node :body)))))
+      (jsdoc--infer-type (treesit-node-child-by-field-name node "body")))))
 
 (defun jsdoc--get-throw-type (node)
```

Comment:
I wasn't able to suggest edit because it's outside the diff range, but for L214:

```diff
- (--> (jsdoc--get-returned-type-of-statement node 'throw_statement)
+ (--> (jsdoc--get-returned-type-of-statement node "throw_statement")
```